File: src/code/python/data_structures/graph.py
class Graph:
    def __init__(self) -> None:
        self.graph = {}

    def add_vertex(self, vertex: str) -> None:
        if vertex not in self.graph:
            self.graph[vertex] = []

    def add_edge(self, a: str, b: str) -> None:
        self.add_vertex(a)
        self.add_vertex(b)
        self.graph[a].append(b)
        self.graph[b].append(a)

    def get_neighbors(self, vertex: str) -> list[str]:
        return self.graph.get(vertex, [])

    def __repr__(self) -> str:
        output = ''

        for vertex, neighbors in self.graph.items():
            output += f'{vertex} - {' - '.join(neighbors)}\n'

        return output

File: src/code/ruby/data_structures/graph.rb
class Graph
  attr_accessor :graph

  def initialize
    @graph = {}
  end

  def add_vertex(vertex)
    @graph[vertex] ||= []
  end

  def add_edge(a, b)
    add_vertex(a)
    add_vertex(b)
    @graph[a] << b
    @graph[b] << a
  end

  def get_neighbors(vertex)
    @graph.fetch(vertex, [])
  end

  def to_s
    output = ''
    @graph.each do |vertex, neighbors|
      output += "#{vertex} - #{neighbors.join(' - ')}\n"
    end
    output
  end
end
